x/ecocredit/base/keeper/msg_send.go

```diff
@@ -49,125 +49,212 @@ func (k Keeper) sendEcocredits(sdkCtx sdk.Context, credit *types.MsgSend_SendCre
 	}
 	precision := creditType.Precision
 
-	batchSupply, err := k.stateStore.BatchSupplyTable().Get(ctx, batch.Key)
+	creditDecs, err := utils.GetNonNegativeFixedDecs(precision, credit.TradableAmount, credit.RetiredAmount)
```

Comment:
Refactored `sendEcocredits` to support self-transfer.

Reply:
We should probably still error in basic validation if a user attempts to send credits to themselves but the refactor makes sense either way. We could allow self-transfer with these changes but I think better not to have the user spend gas for what would most likely be a user error.

Reply:
I did a bit of an audit and additional refactoring based off your branch, which could be merged into your branch or considered for a followup. I think it improves upon what you fixed here and makes things a bit more readable and consistent and avoids passing storage as a parameter. https://github.com/regen-network/regen-ledger/pull/1680